src/rvsdg/simplify_branches.rs

```diff
@@ -0,0 +1,623 @@
+            // branch.
+            for (id, edge, val) in branch_meta
+                .branches
+                .iter()
+                .flat_map(|(id, edges)| edges.iter().map(move |(edge, val)| (id, *edge, val)))
+            {
+                let Some(lit) = to_lit(val) else {
+                    continue;
+                };
+                // Count downwards from usize::MAX to avoid collisions with other placeholders
+                let node_bound = usize::MAX - var_counter;
+                var_counter += 1;
+                let (source, target) = self.graph.edge_endpoints(edge).unwrap();
+                let weight = self.graph.remove_edge(edge).unwrap();
+                let block_name = BlockName::Placeholder(node_bound);
+                let mid = self.graph.add_node(BasicBlock::empty(block_name));
+                self.graph.add_edge(source, mid, weight);
+                // NB: We rely on the optimize_direct_jumps pass to collapse this
+                // back down.
+                self.graph.add_edge(
+                    mid,
+                    target,
+                    Branch {
+                        op: BranchOp::Jmp,
+                        pos: None,
+                    },
+                );
+                val_analysis.add_assignment(mid, id.clone(), ValueInfo::Known(lit));
+            }
+            // Step 3: Compute the fixpoint of the value analysis.
+            val_analysis.compute_fixpoint(self);
+            // Step 4: Rewrite branches:
+            // * For each administrative node `n``...
+            // * For each outgoing branch [edge e1] with cond val `v` for `id`
+            // * Check if `id` was written to in `n`, if it was, then move on
+            // _unless_ we know the value of `id`; in which case we can replace the branch with a jump.
+            // * Otherwise, check if a predecessor [via edge e2] node has `v` as a
+            // known value for `id`.
+            // * If so, copy the contents of the admin node to that predecessor, and
+            // reroute e2 to the target of e1.
+            let mut scratch = Vec::new();
+            for admin_node in &branch_meta.admin_nodes {
+                let mut walker = self
+                    .graph
+                    .neighbors_directed(*admin_node, Direction::Outgoing)
+                    .detach();
+                // Don't reroute past the exit node. We want to make sure it stays reachable.
+                if admin_node == &self.exit {
+                    continue;
+                }
+                while let Some((outgoing, succ)) = walker.next(&self.graph) {
+                    let BranchOp::Cond { arg, val, .. } = self.graph[outgoing].op.clone() else {
+                        continue;
+                    };
+                    let Some(val) = to_lit(&val) else {
+                        continue;
+                    };
+                    if val_analysis.data[admin_node].kills.contains(&arg) {
```

Comment:
I think this `if` is a special, easy case- can it be broken out into a helper? Or a whole different pass perhaps?

Reply:
The thing that makes that harder is the `break` vs. `continue` case. We can lift that into its own type, but that + a bunch of params creates a bit of an odd helper. I think this might be better but I'm not sure.

As for a separate pass I think that'd require that we redo a bunch of the scaffolding for the while loops? Not sure if that's worth it either.

```rust
    enum Next { Break, Continue }
    fn fold_constant_branch(
        &mut self,
        admin_node: NodeIndex,
        succ: NodeIndex,
        arg: &Identifier,
        val_analysis: &ValueAnalysis,
        val: &Literal,
    ) -> (Next, bool) {
        if succ != self.exit && self.graph.neighbors(admin_node).any(|x| x == self.exit) {
            // Don't remove any outgoing links to the exit node.
            return (Next::Break, false);
        }
        // We assign to the branched-on argument in the admin
        // node. See if we can fold the constant branch here.
        let ValueInfo::Known(lit) = val_analysis.data[&admin_node].get_output(arg) else {
            return (Next::Continue, false);
        };
        if &lit != val {
            return (Next::Continue, false);
        }
        // okay, we have found a matching edge. Replace this branch
        // with a jump.
        let mut walker = self
            .graph
            .neighbors_directed(admin_node, Direction::Outgoing)
            .detach();
        while let Some((outgoing, _)) = walker.next(&self.graph) {
            self.graph.remove_edge(outgoing);
        }
        self.graph.add_edge(
            admin_node,
            succ,
            Branch {
                op: BranchOp::Jmp,
                pos: None,
            },
        );
        // Don't run the rest of the inner loop.
        (Next::Break, true)
}
```

Reply:
I agree the helper is weird
I still think a separate while loop may be cleaner but I'm not hung up on this